content/courses/intro-to-embedded-sys-prog/chapters/handling_interrupts.rst

```diff
@@ -6,177 +6,1537 @@ Handling Interrupts
+triggered at the same time, to define the order in which the interrupts
+are presented and the handlers invoked. The canonical model is that only
+higher-priority interrupts can preempt handlers executing in response to
+interrupts with lower or equal priority.
+
+Ada defines a model for hardware interrupts and interrupt handling that
+closely adheres to the conceptual model described above. If you have
+experience with interrupt handling, you will recognize them in the Ada
+model. One very important point to make about the Ada facilities is that
+they are highly portable, so they don't require extensive changes when
+moving to an new target computer. Part of that portability is due to the
+language-defined model.
+
+Before we go into the Ada facility details, there's a final point.
+Sometimes we *do* want the application to wait for the external device.
+When would that be the case? To answer that, we need to introduce another
+term. The act of saving and restoring the state of the interrupted
+application software is known as "interrupt context switching." If the
+time for the device to complete the application request is approximately
 that of the context switching, the application might as well wait for
-the device.
+the device after issuing the request.
+
+Another reason to consider polling is that the architectural complexity
+of interrupt handling is greater than that of polling. If your system
+has some number of devices to control and polling them would be fast
+enough for the application to meet requirements, it is simpler to do so.
+But that will likely only work for a few devices, or at least a few that
+have short response time requirements.
 
-To implement the waiting, the application code can simply enter a loop,
+The application code can wait for the device by simply entering a loop,
 exiting only when some external device status flag indicates completion
-of the function. (You could have an upper bound on the loop, or a
-timeout, if necessary.) That active waiting, via looping, is known as
-"polling." As mentioned earlier, polling only makes sense for very fast
-device interactions. That's a comparatively rare situation though, so
-polling should not be your default design assumption. Besides, active
-polling consumes power. On an embedded platform, conserving power is
-often important.
+of the function. The loop itself, in its simplest form, would contain
+only the test for exiting. As mentioned earlier, polling in a tight loop
+like this only makes sense for very fast device interactions. That's not
+the usual situation though, so polling should not be your default design
+assumption. Besides, active polling consumes power. On an embedded
+platform, conserving power is often important.
+
+That polling loop that only exits when the device indicates completion
+will be an "infinite" (non-terminating) loop if the device can fail to
+do so. Or maybe it might take too long in some odd case. If you don't
```

Comment:
This sentence could be confusing for the reader.

Reply:
Agreed, and updated.